File: course1/week1/merge.js
// Merge sort implementation in JS
const sort = (array) => {

  // Base case
  if (array.length <= 1) {
    return array;
  }
  
  const mid = Math.floor(array.length / 2);
  const left = array.slice(0, mid);
  const right = array.slice(mid);

  return merge(sort(left), sort(right));
}

const merge = (left, right) => {
  let il = 0, ir = 0;
  const result = [];

  // Merge
  while (il < left.length && ir < right.length) {
    if (left[il] < right[ir]) {
      result.push(left[il]);
      il++;
    }
    else {
      result.push(right[ir]);
      ir++;
    }
  }

  // End cases
  while (il < left.length) {
    result.push(left[il]);
    il++;
  }
  while (ir < right.length) {
    result.push(right[ir]);
    ir++;
  }

  return result;
}

console.log(sort([5, 234, 3, 3, 4, 30, 1, 9, 39]));

File: course1/week1/merge.rb
def sort(ary)
  len = ary.length
  return ary if len <= 1
    
  mid = len / 2
  left = ary[0, mid]
  right = ary[mid..-1]

  return merge(sort(left), sort(right))
end

def merge(left, right)
  li = 0
  ri = 0
  result = []

  while li < left.length && ri < right.length
    if left[li] < right[ri]
      result << left[li]
      li += 1
    else
      result << right[ri]
      ri += 1
    end
  end

  while li < left.length
    result << left[li]
    li += 1
  end

  while ri < right.length
    result << right[ri]
    ri += 1
  end

  result
end

p sort([8, 3, 4, 2, 1, 7, 9, 10])
